src/lib.rs

```diff
@@ -0,0 +1,189 @@
+use hyper::{Request, Response, Body};
+use hyper::service::{make_service_fn, service_fn};
+use pam::constants::{PamFlag, PamResultCode, PAM_TEXT_INFO, PAM_ERROR_MSG};
+use pam::conv::Conv;
+use pam::items::RHost;
+use pam::module::{PamHandle, PamHooks};
+use tokio::sync::Notify;
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::time::Duration;
+use pam::pam_try;
+use tokio::time::sleep;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use std::convert::Infallible;
+use tokio::runtime::Runtime;
+use std::net::TcpListener;
+
+struct PamDiscord;
+pam::pam_hooks!(PamDiscord);
+
+const API_ENDPOINT: &str = "https://discord.com/api";
+
+impl PamHooks for PamDiscord {
+    // This function performs the task of authenticating the user.
+    fn sm_authenticate(pamh: &mut PamHandle, args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode { 
+        println!("test");
+        let args: Vec<_> = args
+            .iter()
+            .map(|s| s.to_string_lossy())
+            .collect();
+        let args: HashMap<&str, &str> = args
+            .iter()
+            .map(|s| {
+                let mut parts = s.splitn(2, '=');
+                (parts.next().unwrap(), parts.next().unwrap_or(""))
+            })
+            .collect();
+        let client_id: &str = match args.get("client_id") {
+            Some(client_id) => client_id,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let client_secret: &str = match args.get("client_secret") {
+            Some(client_secret) => client_secret,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let guild_id: &str = match args.get("guild") {
+            Some(guild) => guild,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let role: &str = match args.get("role") {
+            Some(role) => role,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let conv = match pamh.get_item::<Conv>() {
+            Ok(Some(conv)) => conv,
+            Ok(None) => {
+                unreachable!("No conv available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_conv");
+                return err;
+            }
+        };
+        let state: &&str = &"unknown";
+        let rhost = match pamh.get_item::<RHost>() {
+            Ok(Some(rhost)) => rhost,
+            Ok(None) => {
+                unreachable!("No rhost available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_rhost");
```

Comment:
We can't return Rust errors since this is inside PAM, so printing makes sense, but we should use `eprintln!` instead, since that prints to stderr.